File: dynamic_programming/two_setsII/two_sets.py
from sys import stdin

MOD = 10 ** 9 + 7

# Knapsack problem
# Can reduce space from O(n^3) to O(n^2) using knapsack technique
# TLE in CSES
def two_sets(n):
    s = n * (n + 1) // 2
    if s % 2 != 0:
        return 0
    target = s // 2
    dp = [[0 for _ in range(target + 1)] for _ in range(n + 1)]
    dp[0][0] = 1 
    for i in range(1, len(dp)):
        dp[i][0] = 1
        for j in range(1, len(dp[0])):
            dp[i][j] = dp[i - 1][j] + (dp[i - 1][j - i] if j - i >= 0 else 0)
    return (dp[n][target] // 2) % MOD

def read_num():
    return int(stdin.readline())

def read_nums():
    return [int(num) for num in stdin.readline().split()]

if __name__ == '__main__':
    print(two_sets(read_num()))

File: dynamic_programming/two_setsII/two_sets.cpp
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

// knapsack problem 
// mod arithmetic rules to prevent overflow are a bit tricky

const int MOD = pow(10, 9) + 7;

int num_two_sets(int n) {
    int s = (n + 1) * n / 2;
    if (s & 1) {
        return 0;
    }
    int desired_sum = s / 2;
    vector<long long> dp(desired_sum + 1);
    dp[0] = 1;

    for (int i = 1; i <= n; i++) {
        for (int j = desired_sum - i; j >= 0; j--) {
            dp[j + i] += dp[j];
            dp[j + i] %= 2 * MOD;
        }
    }
    return dp[desired_sum] / 2;
}

vector<int> read_nums(int size) {
    vector<int> nums(size);
    for (int i = 0; i < size; i++) {
        int num;
        scanf("%d", &num);
        nums[i] = num;
    }
    return nums;
}

int read_num() {
    int num;
    scanf("%d", &num);
    return num;
}

int main() {
    cout << num_two_sets(read_num()) << endl;
}